### PR TITLE
task_checker(dm): refactor and expose an auto resume function

### PR DESCRIPTION
**Pull request by lance6716**
Signed-off-by: lance6716 <lance6716@gmail.com>

### What problem does this PR solve?

Issue Number: ref #4287

### What is changed and how it works?

as title. For dataflow engine, an unit can be checked auto resume by build an `AutoResumeTimes` and `UpdateResumeStrategy`

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

Side effects

Related changes

### Release note 

```release-note
`None`.
```

### CODE REVIEW

File: dm/dm/worker/task_checker.go [hunk 1]
```diff
@@ -104,33 +104,13 @@ type TaskStatusChecker interface {
...
-func newBackoffController() *backoffController {
-	return &backoffController{
-		backoffs:         make(map[string]*backoff.Backoff),
-		latestPausedTime: make(map[string]time.Time),
-		latestBlockTime:  make(map[string]time.Time),
-		latestResumeTime: make(map[string]time.Time),
-	}
+// AutoResumeTimes contains some Time and Backoff that are related to auto resume.
+// This structure is exposed for DM as library.
+type AutoResumeTimes struct {
```
Comment (D3Hunter) [inline dm/dm/worker/task_checker.go:109]: the suffix `Times` is confusing to me, especially used in variables like `subtaskTimes`, `relayTimes`, cannot see it's related to auto resume, may be `AutoResume[Info]` `subtaskAutoResume` `relayAutoResume`?

File: dm/dm/worker/task_checker.go [hunk 2]
```diff
@@ -254,7 +236,19 @@ func isResumableError(err *pb.ProcessError) bool {
 	return true
 }
 
-func (tsc *realTaskStatusChecker) getResumeStrategy(stStatus *pb.SubTaskStatus, duration time.Duration) ResumeStrategy {
+// UpdateResumeStrategy updates times and returns ResumeStrategy for a subtask.
+// When ResumeDispatch and the subtask is successfully resumed, LatestResumeTime
+// and backoff should be updated.
+// This function is exposed for DM as library.
+func UpdateResumeStrategy(
```
Comment (D3Hunter) [inline dm/dm/worker/task_checker.go:243]: maybe make those functions methods of `AutoResumeTimes`

**Review by D3Hunter — APPROVED** [review]

**Review by Ehco1996 — APPROVED** [review]
LGTM

**Comment by GMHDBJD** [comment]
Can we make AutoResumeInfo as a default interface, and TaskAutoResume/RelayAutoResume implement the CheckResume and Resume func? just like the template method pattern

**Comment by lance6716** [comment]
> Can we make AutoResumeInfo as a default interface, and TaskAutoResume/RelayAutoResume implement the CheckResume and Resume func? just like the template method pattern

left for future work🙈

**Pull request merged**